text/0000-native-web-component-support.md

```diff
@@ -0,0 +1,143 @@
+supported.
+
+```js
+customElements.define('hello-world', class HelloWorld {});
+```
+
+```html
+<template>
+  <hello-world></hello-world>
+</template>
+```
+
+This is because, by design, LWC relies on a custom module loader and never defers to the custom
+element registry when resolving components. The Salesforce platform is based on a multi-tenant
+architecture, and as such, it relies on a directory-based module resolution system that is
+compatible with how developers create components on the platform. This basically guarantees that no
+two parties can define the same component, allowing LWC to side-step a variety of security concerns.
+
+This proposal is about adding native web component support for off-platform users, with the
+additional goal of exploring how we might do the same for platform users. Any solution should be
+forward-compatible with [Scoped Custom Element Registries].
+
+## Motivation
+
+The long term goal of LWC is to provide a thin abstraction layer on top of native browser APIs.
+Adding support for the custom element registry for off-platform applications would be a step towards
+that goal.
+
+For platform applications, the major motivation is to facilitate platform integration and allow
+developers to migrate existing tooling and solutions without having to utilize LWC.
+
+## Detailed design
+
+### OSS restriction
+
+In order for LWC to support native web components for OSS applications, the following needs to
+happen.
+
+1. Prevent LWC components from using the same custom element as a native web component.
+2. Prevent native web components from using the same custom element as an LWC component.
+3. Remove the restriction that LWC must know about all available components before rendering.
+
+In order to prevent namespace collisions between LWC components and native web components, LWC
+components should be registered in the same custom element registry that native web components will
+be registered in.
+
+The third requirement of relaxing the restriction that LWC must know about all available components
+before rendering, would be a significant change. Currently, if an unknown component is encountered,
+an exception is thrown. This behavior should be modified so that unrecognized custom elements are
+rendered as-is with the expectation that they will be natively upgraded. This opens up the
+possibility of rendering custom elements that are `HTMLUnknownElement`, which is basically an
+extension of `HTMLElement` without any additional fields or methods.
+
+To ensure that only off-platform applications can use this feature, it should be implemented behind
+a feature flag (e.g., `ENABLE_CUSTOM_ELEMENT_REGISTRY_RESOLUTION`).
+
+### Platform restriction
+
+In addition to the OSS restrictions outlined above, in order for LWC to support native web
+comopnents for platform applications, the following needs to happen.
+
+1. Prevent registration of custom elements in a namespace different from the current one.

```

Comment:
This restriction is too aggressive IMO. This will prevent platform developers to reuse existing web component libraries and greatly reduce the usefulness of this proposal.

I also don't think LWC should be in the business of restricting custom elements from being registered based on the current namespace. Locker should be in charge of this.

Reply:
> I take it you think Locker should hijack the global `customElements.define()` and handle the namespacing there?

Correct

Reply:
Assuming that other frameworks are able to handle custom elements at runtime (props/attrs/events/etc.) without intercepting the `define()` process, my assumption is that LWC should be able to do the same.